### PR TITLE
feat (Data/Set/Basic): add lemmas about pairs

### PR DESCRIPTION
**Pull request by mehatamm**
Previously, when a set was a subset of a pair, or a pair was a subset of a set, being able to say something easily about those sets required going into the insertion api. We add lemmas about those cases, in addition to lemmas about taking the difference of a pair by one of its members and a useful iff statement for when a set is a subset of an insertion. We move the sections of lemmas about pairs further down in the file as it now depends on lemmas earlier in the file.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Set/Basic.lean [hunk 1]
```diff
@@ -2081,6 +2059,16 @@ theorem diff_singleton_sSubset {s : Set α} {a : α} : s \ {a} ⊂ s ↔ a ∈ s
...
 
+theorem subset_insert_iff {s t : Set α} {x : α} :
+    s ⊆ insert x t ↔ s ⊆ t ∨ (x ∈ s ∧ s \ {x} ⊆ t) := by
+  rw [← diff_singleton_subset_iff]
+  obtain (hx | hx) := em (x ∈ s)
+  · rw [and_iff_right hx]
+    exact ⟨fun h ↦ Or.inr h, fun h ↦ h.elim (fun hst ↦ (diff_subset _ _).trans hst) id⟩
+  rw [diff_singleton_eq_self hx]
+  tauto
+#align set.subset_insert_iff Set.subset_insert_iff
```
Comment (urkud) [inline Mathlib/Data/Set/Basic.lean:2070]: There is no `set.subset_insert_iff` in mathlib3, so no need to `#align`.

File: Mathlib/Data/Set/Basic.lean [hunk 2]
```diff
@@ -2081,6 +2059,16 @@ theorem diff_singleton_sSubset {s : Set α} {a : α} : s \ {a} ⊂ s ↔ a ∈ s
   sdiff_le.lt_iff_ne.trans <| sdiff_eq_left.not.trans <| by simp
 #align set.diff_singleton_ssubset Set.diff_singleton_sSubset
 
+theorem subset_insert_iff {s t : Set α} {x : α} :
```
Comment (urkud) [inline Mathlib/Data/Set/Basic.lean:2062]: The `Finset` version aligns with `diff_singleton_subset_iff.symm`, not with this lemma.

Comment (mehatamm) [reply Mathlib/Data/Set/Basic.lean:2062]: Hi, I'm not sure what you mean by this. What do I need to do to fix this?

Comment (winstonyin) [reply Mathlib/Data/Set/Basic.lean:2062]: Sorry for a slow reply there. You're naming this lemma `Set.subset_insert_iff`, but the lemma statement looks quite different than the `Finset` counterpart `Finset.subset_insert_iff`. In fact, the latter corresponds to the symm of `Set.diff_singleton_subset_iff`. I suggest that you rename this lemma something different to reduce user surprise. I don't have a great name in mind, maybe something like `subset_insert_iff_subset_or_mem_and_diff_subset` but that's quite a mouthful.

File: Mathlib/Data/Set/Basic.lean [hunk 3]
```diff
@@ -2158,6 +2145,53 @@ theorem subset_symmDiff_union_symmDiff_right (h : Disjoint t u) : s ⊆ s ∆ t
...
+#align set.pair_eq_pair_iff Set.pair_eq_pair_iff
+
+theorem pair_diff_left {x y : α} (hne : x ≠ y) : ({x, y} : Set α) \ {x} = {y} := by
+  rw [insert_diff_of_mem _ (by exact rfl : x ∈ {x}), diff_singleton_eq_self (by simpa)]
+
+theorem pair_diff_right {x y : α} (hne : x ≠ y) : ({x, y} : Set α) \ {y} = {x} := by
+  rw [pair_comm, pair_diff_left hne.symm]
+
+@[simp]
+theorem pair_subset_iff {x y : α} {s : Set α} : {x,y} ⊆ s ↔ x ∈ s ∧ y ∈ s := by
```
Comment (winstonyin) [inline Mathlib/Data/Set/Basic.lean:2177]: Minor style comment: please insert a space after commas in the pair notation, i.e. `{x, y}`, not `{x,y}`. Same goes for the rest of the file.

File: Mathlib/Data/Set/Basic.lean [hunk 4]
```diff
@@ -2158,6 +2145,53 @@ theorem subset_symmDiff_union_symmDiff_right (h : Disjoint t u) : s ⊆ s ∆ t
...
+theorem pair_subset {x y : α} {s : Set α} (hx : x ∈ s) (hy : y ∈ s) : {x,y} ⊆ s :=
+  pair_subset_iff.2 ⟨hx,hy⟩
+
+theorem Nonempty.subset_pair_iff {x y : α} {s : Set α} (hs : s.Nonempty) :
+    s ⊆ {x,y} ↔ s = {x} ∨ s = {y} ∨ s = {x,y} := by
+  obtain (rfl | hne) := eq_or_ne x y
+  · rw [pair_eq_singleton, hs.subset_singleton_iff]; simp
+  rw [subset_insert_iff, subset_singleton_iff_eq, subset_singleton_iff_eq, diff_eq_empty,
+    iff_false_intro hs.ne_empty, false_or, and_or_left, ← singleton_subset_iff,
+    ← subset_antisymm_iff, eq_comm (b := s), ←or_assoc, or_comm (a := s = _), or_assoc]
```
Comment (winstonyin) [inline Mathlib/Data/Set/Basic.lean:2189]: Add a space after `←`. Linter is complaining about this.

**Comment by YaelDillies** [comment]
Closing due to inactivity. Please reopen if you want to revive it!

**Pull request closed without merging**